main/say.c

```diff
@@ -353,26 +354,35 @@ static int say_digit_str_full(struct ast_channel *chan, const char *str, const c
 static struct ast_str* ast_get_money_en_dollars_str(const char *str, const char *lang)
 {
 	const char *fnr;
-
-	double dollars = 0;
-	int amt, cents;
+	const char *period;
+	int amt, dollars, cents = 0;
 	struct ast_str *fnrecurse = NULL;
+	struct ast_str *filenames;
 
-	struct ast_str *filenames = ast_str_create(20);
+	if (ast_strlen_zero(str)) {
+		return NULL;
+	}
+
+	filenames = ast_str_create(20);
 	if (!filenames) {
 		return NULL;
 	}
 	ast_str_reset(filenames);
 
-	if (sscanf(str, "%30lf", &dollars) != 1) {
-		amt = 0;
-	} else { /* convert everything to cents */
-		amt = dollars * 100;
+	/* Don't use sscanf because floating point rounding
+	 * could distort the cents units. Just parse as string. */
+	dollars = atoi(str); /* atoi will stop at the period, if there is one */
+	period = strchr(str, '.');
+	if (period) {
+		/* Cents are after the decimal point, if present */
+		period++;
+		if (!ast_strlen_zero(period)) {
+			cents = atoi(period);
```

Comment:
I think it'll be incorrect if the amount string has more then 2 digits after the period,
for example SAYFILES(2.051,money)

Reply:
Good catch, 1 digit was also not handled correctly